src/neurostatslib/basis.py

```diff
@@ -533,6 +537,218 @@ def _check_n_basis_min(self) -> None:
+        Evaluate the B-spline basis functions with given sample points.
+    """
+
+    def __init__(self, n_basis_funcs: int, order: int = 2):
+        super().__init__(n_basis_funcs, order=order)
+        if self.order < 2:
+            raise ValueError(
+                f"Order >= 2 required for cyclic B-spline, "
+                f"order {self.order} specified instead!"
+            )
+
+    def _evaluate(self, sample_pts: NDArray) -> NDArray:
+        """
+        Evaluate the B-spline basis functions with given sample points.
+
+        Parameters
+        ----------
+        sample_pts :
+            The sample points at which the B-spline is evaluated. Must be a tuple of length 1.
+
+        Returns
+        -------
+        NDArray
+            The basis function evaluated at the samples (Time points x number of basis)
+
+        Raises
+        ------
+        AssertionError
+            If the sample points are not within the B-spline knots range unless `outer_ok=True`.
+
+        Notes
+        -----
+        This method evaluates the B-spline basis functions at the given sample points.
+        It requires the knots to be defined through the `_generate_knots` method.
+
+        The evaluation is performed by looping over each element and using `splev` from
+        SciPy to compute the basis values.
+        """
+        super()._check_samples_non_empty(sample_pts)
+
+        self._generate_knots(sample_pts, 0.0, 1.0, is_cyclic=True)
+
+        # for cyclic, do not repeat knots
+        self.knot_locs = np.unique(self.knot_locs)
+
+        knots_orig = self.knot_locs.copy()
+
+        nk = knots_orig.shape[0]
+
+        # make sure knots are sorted
+        knots_orig.sort()
+
+        # extend knots
+        xc = knots_orig[nk - 2 * self.order + 1]
+        knots = np.hstack(
+            (
+                self.knot_locs[0]
+                - self.knot_locs[-1]
+                + self.knot_locs[nk - self.order : nk - 1],
+                self.knot_locs,
+            )
+        )
+        ind = sample_pts > xc
+
+        # temporarily set the extended knots as attribute
+        self.knot_locs = knots
+
+        basis_eval = bspline(
+            sample_pts, self.knot_locs, order=self.order, der=0, outer_ok=True
+        )
+        sample_pts[ind] = sample_pts[ind] - knots.max() + knots_orig[0]
+
+        if np.sum(ind):
+            basis_eval[ind] = basis_eval[ind] + bspline(
+                sample_pts[ind], self.knot_locs, order=self.order, outer_ok=True, der=0
+            )
+        # restore points
+        sample_pts[ind] = sample_pts[ind] + knots.max() - knots_orig[0]
+        # restore the original knots
+        self.knot_locs = knots_orig
+
+        return basis_eval
+
+    def _check_n_basis_min(self) -> None:
+        """Check that the user required enough basis elements.
+
+        Check that Cuclic-BSplineBasis has at least as many basis as the order of the spline +2
+        and at least 2*order - 2 basis.
```

Comment:
Might be worth explaining why this is the case (or including a reference that explains it).

Reply:
I checked my algorithm and I was wrapping too many knots. one need to wrap only order-1 knots instead I was wrapping more. this has been fixed now and it turns out that all the splines has the same condition for the number of basis